### PR TITLE
remove redisDB dependency

### PR DESCRIPTION
**Pull request by nikooo777**
This PR aims at removing the local redisdb dependency that makes it impossible to redistribute work across different servers.

Further work has been put in, here's a summary:
- Use channel IDs while publishing instead of channel names
- Rename API endpoints and associated functions
- Fix indexes in information sent to slack (those +1s that you see in the loop)
- Slightly changed the way the number of "published" claims is counted (making use of the new synced videos list)
- Add 5% slack on UTXOs to avoid waiting a full block even if just a few transactions were made
- Speed up startup by eliminating unnecessary block waits
- Remove waiting for UTXOs by listing UTXOs (instead we just wait for a new block)
- Small bug fixes

### CODE REVIEW

File: ytsync/manager.go [hunk 1]
```diff
@@ -149,7 +165,7 @@ func (s SyncManager) MarkVideoStatus(channelID string, videoID string, status st
 	res, _ := http.PostForm(endpoint, vals)
 	defer res.Body.Close()
 	body, _ := ioutil.ReadAll(res.Body)
-	var response apiSyncUpdateResponse
+	var response apiVideoStatusResponse
```
Comment (lyoshenka) [inline ytsync/manager.go]: you don't have to create a named struct for this if its only used in one place. you can inline it as

```
var response struct {
	Success bool        `json:"success"`
	Error   null.String `json:"error"`
	Data    null.String `json:"data"`
}
```

File: ytsync/ytsync.go [hunk 2]
```diff
@@ -65,12 +65,24 @@ type Sync struct {
 	claimAddress   string
 	videoDirectory string
 	db             *redisdb.DB
+	syncedVideos   map[string]syncedVideo
+	grp            *stop.Group
+	lbryChannelID  string
+
+	videosMapMux sync.Mutex
+	mux          sync.Mutex
```
Comment (lyoshenka) [inline ytsync/ytsync.go]: if you have multiple mutexes, you can't call one of them just `mux`. its not clear what that's for. it could be something like `walletSetupMux`. though once you name it that, it becomes clear that something might be wrong with wrapping the whole wallet setup in a single mutex. does the whole thing *actually* need to be locked?

the mutexes should be pointers, so new copies are not created if the Sync struct is copied

its clearer if you put the mutex right above the thing the mutex is locking, and leave newlines on either side. so in this example, upt `videosMapMux` right above `syncedVideos`. and then maybe rename it to `syncedVideosMux`, like so:

```
...other vars...

syncedVideosMux *sync.Mutex
syncedVideos    map[string]syncedVideo

...other vars...
```

File: ytsync/ytsync.go [hunk 3]
```diff
@@ -65,12 +65,24 @@ type Sync struct {
...
+	wg           sync.WaitGroup
```
Comment (lyoshenka) [inline ytsync/ytsync.go]: this should also be a pointer. but more importantly, you don't need this if you already have a stop.Group. stop.Group works as a combo WaitGroup + channel that will be closed to indicate stopping (and can be closed safely multiple times). so use `grp.Add()` and `grp.Wait()` instead

Comment (lyoshenka) [reply ytsync/ytsync.go]: it won't let me comment below, so I'm commenting here:

you call `Add()` and `Done()` inside `startWorker(workerNum int)`, but the correct pattern is to make those calls outside the function. startWorker() doesn't know if its being run asynchronously or not. you only need a waitgroup if it is. there are also subtle concurrency issues with calling it inside the function. so the right way to go is to remove Add and Done from inside startWorker, and do this:

```
	for i := 0; i < s.ConcurrentVideos; i++ {
		s.wg.Add(1)
		go func() {
			defer s.wg.Done()
			s.startWorker(i)
		}()
	}
```

File: ytsync/ytsync.go [hunk 4]
```diff
@@ -182,6 +195,9 @@ func (s *Sync) FullCycle() (e error) {
 	}
 
 	s.db = redisdb.New()
+	s.videosMapMux.Lock()
```
Comment (lyoshenka) [inline ytsync/ytsync.go]: why is this code all the way down here, so far from line 125 where syncedVideos is first declared?

File: ytsync/ytsync.go (hunk 2 above)
Comment (lyoshenka) [reply ytsync/ytsync.go]: you can use a read/write lock to lock the map for reading when you read it, and writing when you write to it. this lets multiple threads read the data at once, which is safe and blocks less. use `*sync.RWMutex` instead, and call `mux.RLock()` and `mux.RUnlock()` when you're only reading from the variable. leave `mux.Lock()` for writing

File: ytsync/setup.go [hunk 5]
```diff
@@ -44,17 +44,17 @@ func (s *Sync) walletSetup() error {
...
-	numPublished, err := s.daemon.NumClaimsInChannel(s.LbryChannelName)
-	if err != nil {
-		return err
-	}
+	s.videosMapMux.Lock()
+	numPublished := uint64(len(s.syncedVideos)) //should we only count published videos? Credits are allocated even for failed ones...
+	s.videosMapMux.Unlock()
 	log.Debugf("We already published %d videos", numPublished)
 
 	if float64(numOnSource)-float64(numPublished) > float64(s.Manager.VideosLimit) {
```
Comment (lyoshenka) [inline ytsync/setup.go]: why are all these cast to floats? they can be ints here

File: ytsync/ytsync.go [hunk 6]
```diff
@@ -493,10 +510,32 @@ func (s *Sync) processVideo(v video) (err error) {
...
+	sv, ok := s.syncedVideos[v.ID()]
+	s.videosMapMux.Unlock()
+	alreadyPublished := ok && sv.Published
+
+	neverRetryFailures := []string{
+		"Error extracting sts from embedded url response",
+		"the video is too big to sync, skipping for now",
+	}
+	if ok && !sv.Published && util.SubstringInSlice(sv.FailureReason, neverRetryFailures) {
+		log.Println(v.ID() + " can't ever be published")
```
Comment (lyoshenka) [inline ytsync/ytsync.go]: how often does this happen, and do you know why? is it within our control?

Comment (nikooo777) [reply ytsync/ytsync.go]: the first happens due to youtube restrictions (either region, age or deleted videos)
the second happens when we hit the defined size limit (it's in our control).

While we can't do anything about the first, we can raise the limits for the second, but as per several discussions we'll keep it at 2GB or less.

Comment (nikooo777) [reply ytsync/ytsync.go]: out of 100k published videos we have about 1% that failed for those two (leading) reasons: https://scrn.storni.info/2018-08-09_12-24-24-339707393.png

File: ytsync/setup.go (hunk 5 above)
Comment (nikooo777) [reply ytsync/setup.go]: no clue. but you're completely right. Will fix

File: ytsync/ytsync.go (hunk 4 above)
Comment (nikooo777) [reply ytsync/ytsync.go]: the only reason i put it there is because there were a bunch of assignments there and i wanted to keep them grouped, but I agree it's not very read-able this way. I'll move them up

File: ytsync/ytsync.go (hunk 3 above)
Comment (nikooo777) [reply ytsync/ytsync.go]: I need to look into that, this is from your original code and I don't remember ever changing it. Thanks for the pointers there.

File: ytsync/ytsync.go (hunk 2 above)
Comment (nikooo777) [reply ytsync/ytsync.go]: Will rename and move the mutexes. 
The reason I didn't use read/write locks is that i don't want the application to read when it's being written to, plus the locks are held for a very very short time so there would be no noticeable improvement, only a higher risk of race conditions happening during wallet refills.

File: ytsync/manager.go (hunk 1 above)
Comment (nikooo777) [reply ytsync/manager.go]: ok, will look into that

File: ytsync/ytsync.go (hunk 3 above)
Comment (nikooo777) [reply ytsync/ytsync.go]: I looked more into why we need to lock the whole walletSetup function. The reason I'm doing that is to avoid multiple threads from refilling the wallet concurrently. I don't think I can easily break up the the function to lock fewer lines of code. I think it's fair to leave it like that.

I removed the wait group in favor of the stop group

File: ytsync/manager.go [hunk 7]
```diff
@@ -90,13 +90,19 @@ func (s SyncManager) fetchChannels(status string) ([]apiYoutubeChannel, error) {
...
 }
 
-type apiSyncUpdateResponse struct {
-	Success bool        `json:"success"`
-	Error   null.String `json:"error"`
-	Data    null.String `json:"data"`
+type apiChannelStatusResponse struct {
+	Success bool          `json:"success"`
+	Error   null.String   `json:"error"`
+	Data    []syncedVideo `json:"data"`
```
Comment (tiger5226) [inline ytsync/manager.go]: Data? I would suggest `syncedVideos instead of data

File: ytsync/manager.go [hunk 8]
```diff
@@ -149,7 +159,11 @@ func (s SyncManager) MarkVideoStatus(channelID string, videoID string, status st
 	res, _ := http.PostForm(endpoint, vals)
 	defer res.Body.Close()
 	body, _ := ioutil.ReadAll(res.Body)
-	var response apiSyncUpdateResponse
+	var response struct {
```
Comment (tiger5226) [inline ytsync/manager.go]: why are we not using the aligned code in the api package? I would make sure we are making things DRY. `api.Response`

File: ytsync/setup.go [hunk 9]
```diff
@@ -14,8 +14,8 @@ import (
 
 func (s *Sync) walletSetup() error {
 	//prevent unnecessary concurrent execution
-	s.mux.Lock()
-	defer s.mux.Unlock()
+	s.walletMux.Lock()
+	defer s.walletMux.Unlock()
```
Comment (tiger5226) [inline ytsync/setup.go]: why are we using a mutex lock? What is wrong with concurrent execution? A mutex is not the standard way to handle synchronization. It can cause deadlocks. Ideally if you want things to run synchronously from different go routines, you use channels in golang.

File: ytsync/setup.go [hunk 10]
```diff
@@ -30,31 +30,32 @@ func (s *Sync) walletSetup() error {
 	balance := decimal.Decimal(*balanceResp)
 	log.Debugf("Starting balance is %s", balance.String())
 
-	var numOnSource uint64
+	var numOnSource int
```
Comment (tiger5226) [inline ytsync/setup.go]: why change this to an int? It better to be more specific than more general.

File: ytsync/setup.go [hunk 11]
```diff
@@ -30,31 +30,32 @@ func (s *Sync) walletSetup() error {
...
 	if s.LbryChannelName == "@UCBerkeley" {
 		numOnSource = 10104
```
Comment (tiger5226) [inline ytsync/setup.go]: what is this?! not part of the PR but this is not good to have in the code base.

File: ytsync/setup.go [hunk 12]
```diff
@@ -155,28 +148,31 @@ func (s *Sync) ensureEnoughUTXOs() error {
...
-	origin := time.Now()
-	for {
-		r, err := s.daemon.UTXOList()
+func (s *Sync) waitForNewBlock() error {
+	status, err := s.daemon.Status()
+	if err != nil {
+		return err
+	}
+
+	for status.Wallet.Blocks == 0 || status.Wallet.BlocksBehind != 0 {
```
Comment (tiger5226) [inline ytsync/setup.go]: so is this ok? If there is a problem it would be an infinite loop no? Maybe have a `maxit`?

File: ytsync/sources/ucbVideo.go [hunk 13]
```diff
@@ -170,25 +170,23 @@ func (v ucbVideo) saveThumbnail() error {
...
-		Language:     strPtr("en"),
-		ClaimAddress: &claimAddress,
-		Thumbnail:    strPtr("https://berk.ninja/thumbnails/" + v.id),
-		License:      strPtr("see description"),
-	}
-
-	if channelName != "" {
-		options.ChannelName = &channelName
+		Title:         &v.title,
+		Author:        strPtr("UC Berkeley"),
```
Comment (tiger5226) [inline ytsync/sources/ucbVideo.go]: Why is the author UC Berkley? should this function be called publishUCBerkley?

Comment (tiger5226) [reply ytsync/sources/ucbVideo.go]: ahhh,  `ucbVideo` Seems odd to have a struct type for a specific author.

File: ytsync/ytsync.go [hunk 14]
```diff
@@ -61,16 +61,27 @@ type Sync struct {
...
+	walletMux *sync.Mutex
+	queue     chan video
+}
 
-	mux   sync.Mutex
-	wg    sync.WaitGroup
-	queue chan video
+func (s *Sync) AppendSyncedVideo(videoID string, published bool, failureReason string) {
+	s.syncedVideosMux.Lock()
+	defer s.syncedVideosMux.Unlock()
```
Comment (tiger5226) [inline ytsync/ytsync.go]: I don't understand why we have a mutex lock here. Why would this get called twice in different go routines on the same sync instance?

File: ytsync/ytsync.go [hunk 15]
```diff
@@ -110,10 +121,26 @@ func (s *Sync) FullCycle() (e error) {
...
-	err := s.Manager.setChannelSyncStatus(s.YoutubeChannelID, StatusSyncing)
+	s.syncedVideosMux = &sync.Mutex{}
+	s.walletMux = &sync.Mutex{}
+	s.db = redisdb.New()
+	s.grp = stop.New()
+	s.queue = make(chan video)
+	interruptChan := make(chan os.Signal, 1)
+	signal.Notify(interruptChan, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-interruptChan
```
Comment (tiger5226) [inline ytsync/ytsync.go]: we should not do this here. Please use the stopper pattern grin built in the stop package.

File: ytsync/ytsync.go [hunk 16]
```diff
@@ -110,10 +121,26 @@ func (s *Sync) FullCycle() (e error) {
...
+		log.Println("Got interrupt signal, shutting down (if publishing, will shut down after current publish)")
+		s.grp.Stop()
+	}()
+	syncedVideos, err := s.Manager.setChannelStatus(s.YoutubeChannelID, StatusSyncing)
 	if err != nil {
 		return err
 	}
+	s.syncedVideosMux.Lock()
+	s.syncedVideos = syncedVideos
+	s.syncedVideosMux.Unlock()
```
Comment (tiger5226) [inline ytsync/ytsync.go]: Last comment on mutex locks. We should discuss the pattern you are using here, as I am almost positive using channels is a better and more robust solution design.

File: ytsync/ytsync.go (hunk 2 above)
Comment (lyoshenka) [reply ytsync/ytsync.go]: a read/write lock does not allow reading during writing. thats the point of every lock. what it does allow is multiple concurrent reads. when you do RLock, others can RLock at the same time. when you Lock (for writing), no one's allowed to Lock or RLock at the same time.

Comment (nikooo777) [reply ytsync/ytsync.go]: Oh yes, you're right, not sure what was going in my mind. I'll swap that

File: ytsync/setup.go (hunk 9 above)
Comment (nikooo777) [reply ytsync/setup.go]: this is the only place the lock is used and no multiple locks are being held that can cause a deadlock here.
However I would like to discuss channels with you to understand how they could be used here (not sure they can)

File: ytsync/setup.go (hunk 10 above)
Comment (nikooo777) [reply ytsync/setup.go]: Grins previous review outlined a mess with casts here and there to make simple math.
I changed everything to int as it's reasonable for the values they represent.

File: ytsync/setup.go (hunk 11 above)
Comment (nikooo777) [reply ytsync/setup.go]: I can explain the berkeley stuff to you via DM, it's all good as it will be eventually removed from here. Not worth changing now

File: ytsync/ytsync.go (hunk 14 above)
Comment (nikooo777) [reply ytsync/ytsync.go]: videos end concurrently, that's why
